pkg/sentry/fsimpl/devpts/line_discipline.go

```diff
@@ -455,6 +456,80 @@ func (*inputQueueTransformer) transform(l *lineDiscipline, q *queue, buf []byte)
 			l.terminal.replicaKTTY.SignalForegroundProcessGroup(kernel.SignalInfoPriv(linux.SIGQUIT))
 		}
 
+		// In canonical mode, some characters need to be handled specially; for example, backspace.
+		if l.termios.LEnabled(linux.ICANON) {
+			// n_tty.c:n_tty_receive_char_canon
+			c := cBytes[0]
+			c_cc := l.termios.ControlCharacters
+
+			// c == c_cc[linux.VKILL] is also handled by n_tty.c:eraser, but this isn't implemented
+			if c == c_cc[linux.VERASE] || (c == c_cc[linux.VWERASE] && l.termios.LEnabled(linux.IEXTEN)) {
+				// n_tty.c:eraser
+				var kill_type int
+				if c == c_cc[linux.VERASE] {
+					kill_type = linux.VERASE
+				} else {
+					kill_type = linux.VWERASE
+				}
+				seen_alnums := 0
+				for len(q.readBuf) > 0 {
+					// Erase a single possibly multibyte character
+					var to_erase byte
+					cnt := 0
+					is_continuation := true
+					for {
+						to_erase = q.readBuf[len(q.readBuf)-cnt-1]
+						cnt++
+						is_continuation = l.termios.IEnabled(linux.IUTF8) && (to_erase&0xc0) == 0x80
+						if cnt == len(q.readBuf) || !is_continuation {
+							break
+						}
+					}
+					if is_continuation {
+						// do not partially erase
+						break
+					}
+
+					if kill_type == linux.VWERASE {
+						if unicode.IsLetter(rune(to_erase)) || unicode.IsDigit(rune(to_erase)) || to_erase == '_' {
+							seen_alnums++
+						} else if seen_alnums > 0 {
+							break
+						}
+					}
+
+					q.readBuf = q.readBuf[:len(q.readBuf)-cnt]
+					if l.termios.LEnabled(linux.ECHO) {
+						if l.termios.LEnabled(linux.ECHOPRT) {
+							// Not implemented
+						} else if kill_type == linux.VERASE && !l.termios.LEnabled(linux.ECHOE) {
+							// Not implemented
+						} else if to_erase == '\t' {
+							// Not implemented
+						} else {
+							iscntrl := to_erase < 0x20 || to_erase == 0x7f
+							echoctl := l.termios.LEnabled(linux.ECHOCTL)
+							if iscntrl && echoctl {
+								l.outQueue.writeBytes([]byte{'\b', ' ', '\b'}, l)
+							}
+							if !iscntrl || echoctl {
+								l.outQueue.writeBytes([]byte{'\b', ' ', '\b'}, l)
+							}
```

Comment:
Can you comment on what these two `if` cases cover? It's not obvious to me.

Reply:
To be honest, I'm not sure either. I rewrote it like this:

```go
charsToDelete := 1
if isCtrl {
	// echoctl controls how we echo control characters, which also determines how we delete them.
	if echoctl {
		// echoctl echoes control characters as ^X, so we need to erase two characters.
		charsToDelete = 2
	} else {
		// if echoctl is disabled, we don't echo control characters so we don't have to erase anything.
		charsToDelete = 0
	}
}
for i := 0; i < charsToDelete; i++ {
	// This sequence is necessary because some terminal emulators do not support
	// deletion of characters just by moving the cursor over them. The sequence
	// ensures that regardless of the terminal's capabilities, the character is
	// erased visually by being overwritten with a space.
	l.outQueue.writeBytes([]byte{'\b', ' ', '\b'}, l)
}
```

I believe this does the same thing. I made some guesses as to why this code is correct; however, I cannot be sure that my reasoning is correct.